piksi_tools/fileio.py

```diff
@@ -238,27 +239,31 @@ def read(self, filename):
         """
         offset = 0
         chunksize = MAX_PAYLOAD_SIZE - 4
-        closure = {'done': False, 'buf': []}
+        closure = {'mostly_done': False, 'done': False, 'buf': {}, 'pending':set()}
 
         def cb(req, resp):
-            if len(closure['buf']) < req.offset:
-                closure['buf'] += [0] * (req.offset - len(closure['buf']))
-            closure['buf'][req.offset:
-                           req.offset + len(resp.contents)] = resp.contents
+            closure['pending'].remove(req.offset)
+            closure['buf'][req.offset] = resp.contents
             if req.chunk_size != len(resp.contents):
```

Comment:
Is this condition guarenteed to happen ONLY on the last packet of the file?

Reply:
By "this" I mean when the chunk_size isn't same as the response contents
```if req.chunk_size != len(resp.contents):```

Reply:
If we ask for ~250 bytes and get back 0 or <250 for the `open()`/`read()` calls on Linux you're guaranteed that something went "wrong", either the call failed, or you reached the end of the file.  If you read past the end of the file in the fileio daemon we just return an empty buffer because the `seek()` call on the file will have failed.  This always seems to be how we detected the end of the file, the new behavior is the `fileio` module having multiple read packets in-flight through the SelectiveRepeater, so we can potentially attempt to read `WINDOW_SIZE*MAX_PACKET_SIZE` bytes past the end of the file.